### PR TITLE
Implement PageAllocation as a handle into a PagedAttentionCache, allowing publishing and releasing an allocation via handle rather than cache

### LINKED ISSUE #607: Manage page allocations through a PageAllocation object, rather than straight in InferenceExecRequest
To manage the lifecycle of page allocations for an inference request, it may be important to use an interface to encapsulate:
- a list of cached pages
- a list of newly allocated pages 
- boundaries between cached and allocated pages
- 3 operations
  - publish (make pages available in cache)
  - release (make pages eligible for eviction)
  - get_page_list (get the full list of pages for use in a vmfb kernel invocation)
 
```python
class PageAllocation(ABC):
    """
    Abstract base class for page allocations in the cache.
    Subclasses only need to implement the core allocation methods.
    """
    @abstractmethod
    def get_page_list(self) -> List[PageInfo]:
        """Returns the list of pages that were allocated."""
        pass

    @abstractmethod
    def publish_pages(self) -> None:
        """
        Makes pages available to other requests after writing is complete.
        Associates tokens with pages and marks them as ready for reading.
        """
        pass

    @abstractmethod
    def release_pages(self) -> None:
        """
        Releases the allocation's reference to pages.
        Pages become eligible for eviction when their reference count reaches zero.
        """
        pass
```

### PR DESCRIPTION
**Pull request by renxida**
Deinitialization looks wonky for now. Will test extensively to get deinit right once I merge #600

Closes #607

### CODE REVIEW

File: shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py [hunk 1]
```diff
@@ -8,9 +8,91 @@
...
+
+
+class PageAllocation(ABC):
+    """
+    Abstract base class for page allocations in the cache.
+    Subclasses only need to implement the core allocation methods.
+    """
+
+    @abstractmethod
+    def get_page_list(self) -> List[PageInfo]:
```
Comment (stbaione) [inline shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py:34]: Completely opinionated nit, might be nice, in terms of interface, if this was a property:

```python
paged_attention_cache.pages[0:index]
```

File: shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py [hunk 2]
```diff
@@ -57,24 +139,9 @@ def acquire_pages_for_tokens(
...
-        """
-        Given a list of tokens and pages containing KV corresponding to these tokens, make these pages available to other requests.
-
-        Associates the tokens with the pages, and mark them as done writing.
-
-        It is assumed that hereafter, the calling request will not modify these pages, at least not the positions [0:len(tokens)].
-        """
-
-        pass  # the base implementation doesn't cache unfinished requests.
+        n_cached_tokens = 0
```
Comment (stbaione) [inline shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py:145]: Looks like this variable isn't used anymore

File: shortfin/python/shortfin_apps/llm/components/messages.py [hunk 3]
```diff
@@ -52,35 +52,22 @@ def reset(self, phase: InferencePhase):
...
 
     def cache_page_indices(self, max_len: int) -> list[int]:
-        if not self.locked_pages:
+        if not self.allocation:
             return []
-        indices = [p.index for p in self.locked_pages]
-        if len(indices) > max_len:
-            return indices[0:max_len]
-        return indices
+        indices = [p.index for p in self.allocation.get_page_list()]
```
Comment (stbaione) [inline shortfin/python/shortfin_apps/llm/components/messages.py:61]:
```python
return [p.index for p in self.allocation.get_page_list()[:max_len]]
```

Saves a couple iterations to not populate entire list, then only return a subset of the list

**Comment by stbaione** [comment]
A couple small comments on form, but functionally, this looks great. There are a decent bit of changes in critical areas, so some unit tests might be nice to include before merging

File: shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py (hunk 1 above)
Comment (renxida) [reply shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py:34]: i like this. let's do this!

File: shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py (hunk 2 above)
Comment (renxida) [reply shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py]: fixd! ty

File: shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py (hunk 1 above)
Comment (renxida) [reply shortfin/python/shortfin_apps/llm/components/kvcache/base_attention_cache.py]: Done!

File: shortfin/python/shortfin_apps/llm/components/messages.py (hunk 3 above)
Comment (renxida) [reply shortfin/python/shortfin_apps/llm/components/messages.py]: nice catch! implemented.

File: shortfin/tests/apps/llm/components/kvcache/base_attention_cache_test.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+      (5,                       1,      False,  "high_concurrency"),               # Higher concurrency, single page
+      (3,                       2,      False,  "multi_page"),                     # Multiple pages per worker
+      (2,                       3,      False,  "more_pages"),                     # More pages than workers, within capacity
+      (TEST_POOL_CAPACITY,      1,      False,  "max_capacity"),                   # Max capacity single pages
+      (TEST_POOL_CAPACITY // 2, 2,      False,  "max_capacity_multi"),             # Max capacity multiple pages
+      (4,                       3,      True ,  "exceeds_total"),                  # 12 pages needed, exceeds capacity
+      (TEST_POOL_CAPACITY + 1,  1,      True ,  "exceeds_workers"),                # More workers than capacity
+      (TEST_POOL_CAPACITY // 2, 3,      True ,  "exceeds_with_multi"),             # Exceeds capacity with multiple pages
+  ],
+)
```
Comment (renxida) [inline shortfin/tests/apps/llm/components/kvcache/base_attention_cache_test.py:87]: Briefly turn off formatter to preserve table alignment.

**Review by stbaione — APPROVED** [review]
LGTM, really nice tests

**Pull request merged**